File: examples/a2a-multi-agent-demo/python/analysis_agent.py
"""
Analysis Agent - Provides data analysis skills for A2A collaboration.

Demonstrates:
- Agent registration with AIM
- Agent Card with multiple skills
- Request verification
- Skill execution
"""
import os
import sys
from typing import Dict, List, Optional

# Add SDK to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../../sdk/python'))

from aim_sdk import register_agent
from aim_sdk.a2a import A2AClient


class AnalysisAgent:
    """
    Analysis Agent providing sentiment analysis, summarization, and trend detection.

    This agent registers with AIM and exposes skills that other agents can discover
    and invoke through the A2A protocol.
    """

    SKILLS = [
        {
            "id": "sentiment-analysis",
            "name": "Sentiment Analysis",
            "description": "Analyze sentiment of text data, returning positive/negative/neutral classification",
            "tags": ["nlp", "sentiment", "text-analysis", "customer-feedback"],
            "inputModes": ["text"],
            "outputModes": ["text"]
        },
        {
            "id": "summarization",
            "name": "Text Summarization",
            "description": "Summarize long text documents into concise summaries",
            "tags": ["nlp", "summarization", "text", "content"],
            "inputModes": ["text"],
            "outputModes": ["text"]
        },
        {
            "id": "trend-detection",
            "name": "Trend Detection",
            "description": "Detect trends and patterns in time-series data",
            "tags": ["analytics", "trends", "data-analysis", "forecasting"],
            "inputModes": ["data"],
            "outputModes": ["data", "text"]
        }
    ]

    def __init__(self, aim_url: str, api_key: str):
        """
        Initialize the Analysis Agent.

        Args:
            aim_url: AIM platform URL
            api_key: API key for authentication
        """
        self.agent_name = "analysis-agent"
        self.aim_url = aim_url

        print(f"[Analysis Agent] Registering with AIM...")

        # Register agent and get AIM client
        self.aim_client = register_agent(
            name=self.agent_name,
            aim_url=aim_url,
            api_key=api_key,
            display_name="Analysis Agent",
            agent_type="ai_agent",
            description="Analysis agent providing sentiment analysis, summarization, and trend detection"
        )

        # Initialize A2A client
        self.a2a = A2AClient(self.aim_client)

        # Store agent ID
        self.agent_id = self.aim_client.agent_id
        print(f"[Analysis Agent] Registered with ID: {self.agent_id[:8]}...")

    def register_agent_card(self) -> Optional[Dict]:
        """
        Register the A2A Agent Card with skills.

        Returns:
            Registered card info or None on failure
        """
        print(f"[Analysis Agent] Registering Agent Card with {len(self.SKILLS)} skills...")

        try:
            # Register skills
            for skill in self.SKILLS:
                self.a2a.register_skill(skill)
                print(f"  - Registered skill: {skill['name']}")

            # Get the registered card
            card = self.a2a.get_agent_card()
            print(f"[Analysis Agent] Agent Card registered successfully")
            return card

        except Exception as e:
            print(f"[Analysis Agent] Card registration note: {e}")
            return None

    def verify_incoming_request(
        self,
        agent_id: str,
        timestamp: int,
        nonce: str,
        signature: str,
        request_hash: str
    ) -> Dict:
        """
        Verify an incoming A2A request signature.

        Args:
            agent_id: Requesting agent's ID
            timestamp: Request timestamp
            nonce: Request nonce
            signature: Ed25519 signature
            request_hash: Hash of request body

        Returns:
            Verification result with agent info and trust score
        """
        print(f"[Analysis Agent] Verifying request from {agent_id[:8]}...")

        result = self.a2a.verify_request(
            agent_id=agent_id,
            timestamp=timestamp,
            nonce=nonce,
            signature=signature,
            request_hash=request_hash
        )

        if result.get("valid"):
            print(f"[Analysis Agent] Request verified! Agent: {result.get('agentName')}")
            print(f"[Analysis Agent] Trust score: {result.get('trustScore', 'N/A')}")
        else:
            print(f"[Analysis Agent] Verification FAILED: {result.get('error')}")

        return result

    def execute_skill(self, skill_id: str, input_data: Dict) -> Dict:
        """
        Execute a skill and return results.

        Args:
            skill_id: ID of skill to execute
            input_data: Input data for the skill

        Returns:
            Skill execution result
        """
        print(f"[Analysis Agent] Executing skill: {skill_id}")

        if skill_id == "sentiment-analysis":
            return self._analyze_sentiment(input_data.get("text", ""))
        elif skill_id == "summarization":
            return self._summarize(input_data.get("text", ""))
        elif skill_id == "trend-detection":
            return self._detect_trends(input_data.get("data", []))
        else:
            return {"error": f"Unknown skill: {skill_id}"}

    def _analyze_sentiment(self, text: str) -> Dict:
        """
        Analyze sentiment of text.

        In production, this would use an NLP model.
        """
        # Simple keyword-based sentiment (demo purposes)
        positive_words = ["excellent", "great", "amazing", "good", "outstanding", "recommend", "happy", "love"]
        negative_words = ["bad", "terrible", "awful", "disappointed", "not satisfied", "poor", "hate"]

        text_lower = text.lower()
        positive_count = sum(1 for word in positive_words if word in text_lower)
        negative_count = sum(1 for word in negative_words if word in text_lower)

        if positive_count > negative_count:
            sentiment = "positive"
            confidence = min(0.5 + (positive_count * 0.1), 0.95)
        elif negative_count > positive_count:
            sentiment = "negative"
            confidence = min(0.5 + (negative_count * 0.1), 0.95)
        else:
            sentiment = "neutral"
            confidence = 0.6

        return {
            "sentiment": sentiment,
            "confidence": round(confidence, 2),
            "wordCount": len(text.split()),
            "positiveIndicators": positive_count,
            "negativeIndicators": negative_count
        }

    def _summarize(self, text: str) -> Dict:
        """
        Summarize text content.

        In production, this would use a summarization model.
        """
        sentences = [s.strip() for s in text.split('.') if s.strip()]

        # Take first 2 sentences as summary (demo purposes)
        summary_sentences = sentences[:2] if len(sentences) > 2 else sentences
        summary = '. '.join(summary_sentences) + '.' if summary_sentences else text

        return {
            "summary": summary,
            "originalLength": len(text),
            "summaryLength": len(summary),
            "compressionRatio": round(len(summary) / max(len(text), 1), 2),
            "sentenceCount": len(sentences)
        }

    def _detect_trends(self, data: List) -> Dict:
        """
        Detect trends in time-series data.

        In production, this would use statistical analysis.
        """
        if not data or len(data) < 2:
            return {
                "trend": "insufficient_data",
                "dataPoints": len(data) if data else 0,
                "confidence": 0.0
            }

        # Extract values
        values = [d.get("value", 0) for d in data if isinstance(d, dict)]

        if len(values) < 2:
            return {
                "trend": "insufficient_data",
                "dataPoints": len(values),
                "confidence": 0.0
            }

        # Simple trend detection
        first_half = sum(values[:len(values)//2]) / (len(values)//2)
        second_half = sum(values[len(values)//2:]) / (len(values) - len(values)//2)

        change_pct = ((second_half - first_half) / first_half) * 100 if first_half != 0 else 0

        if change_pct > 5:
            trend = "upward"
        elif change_pct < -5:
            trend = "downward"
        else:
            trend = "stable"

        return {
            "trend": trend,
            "changePercent": round(change_pct, 1),
            "dataPoints": len(values),
            "startValue": values[0],
            "endValue": values[-1],
            "confidence": 0.78
        }

    def get_skills(self) -> List[Dict]:
        """Return list of available skills."""
        return self.SKILLS

    def get_trust_score(self) -> float:
        """Get the agent's A2A trust score."""
        try:
            score = self.a2a.get_a2a_trust_score()
            return score.a2a_trust_score
        except Exception:
            return 0.0

File: examples/a2a-multi-agent-demo/typescript/src/analysis-agent.ts
/**
 * Analysis Agent - Provides data analysis skills for A2A collaboration.
 */
import { AIMClient, A2AClient, AgentType } from '@opena2a/aim-sdk';

interface SentimentResult {
  sentiment: string;
  confidence: number;
  wordCount: number;
  positiveIndicators: number;
  negativeIndicators: number;
}

interface SummaryResult {
  summary: string;
  originalLength: number;
  summaryLength: number;
  compressionRatio: number;
  sentenceCount: number;
}

interface TrendResult {
  trend: string;
  changePercent?: number;
  dataPoints: number;
  startValue?: number;
  endValue?: number;
  confidence: number;
}

interface SkillDefinition {
  name: string;
  description: string;
  tags: string[];
}

/**
 * Analysis Agent providing sentiment analysis, summarization, and trend detection.
 */
export class AnalysisAgent {
  static readonly SKILLS: SkillDefinition[] = [
    {
      name: 'Sentiment Analysis',
      description: 'Analyze sentiment of text data, returning positive/negative/neutral classification',
      tags: ['nlp', 'sentiment', 'text-analysis', 'customer-feedback'],
    },
    {
      name: 'Text Summarization',
      description: 'Summarize long text documents into concise summaries',
      tags: ['nlp', 'summarization', 'text', 'content'],
    },
    {
      name: 'Trend Detection',
      description: 'Detect trends and patterns in time-series data',
      tags: ['analytics', 'trends', 'data-analysis', 'forecasting'],
    },
  ];

  private readonly agentName = 'analysis-agent';
  private aimClient: AIMClient;
  private a2a!: A2AClient;
  public agentId: string = '';

  constructor(private readonly baseUrl?: string) {
    console.log('[Analysis Agent] Initializing...');

    // SDK handles auth automatically via bundled credentials
    this.aimClient = new AIMClient({
      baseUrl: baseUrl || 'http://localhost:8080',
    });
  }

  async initialize(): Promise<void> {
    console.log('[Analysis Agent] Registering with AIM...');

    const agent = await this.aimClient.registerAgent({
      name: this.agentName,
      displayName: 'Analysis Agent',
      agentType: AgentType.AI_AGENT,
      description: 'Analysis agent providing sentiment analysis, summarization, and trend detection',
    });

    this.agentId = agent.id;
    this.a2a = new A2AClient(this.aimClient, this.agentId);
    console.log(`[Analysis Agent] Registered with ID: ${this.agentId.slice(0, 8)}...`);
  }

  async registerAgentCard(): Promise<Record<string, unknown> | null> {
    console.log(`[Analysis Agent] Registering skills (${AnalysisAgent.SKILLS.length})...`);

    try {
      for (const skill of AnalysisAgent.SKILLS) {
        await this.a2a.registerSkill({
          name: skill.name,
          description: skill.description,
          tags: skill.tags,
        });
        console.log(`  - Registered skill: ${skill.name}`);
      }

      console.log('[Analysis Agent] Skills registered successfully');
      return { skills: AnalysisAgent.SKILLS.length };
    } catch (e) {
      console.log(`[Analysis Agent] Registration note: ${e}`);
      return null;
    }
  }

  executeSkill(
    skillId: string,
    inputData: Record<string, unknown>
  ): SentimentResult | SummaryResult | TrendResult | { error: string } {
    console.log(`[Analysis Agent] Executing skill: ${skillId}`);

    switch (skillId) {
      case 'sentiment-analysis':
        return this.analyzeSentiment((inputData.text as string) || '');
      case 'summarization':
        return this.summarize((inputData.text as string) || '');
      case 'trend-detection':
        return this.detectTrends((inputData.data as Array<Record<string, unknown>>) || []);
      default:
        return { error: `Unknown skill: ${skillId}` };
    }
  }

  private analyzeSentiment(text: string): SentimentResult {
    const positiveWords = ['excellent', 'great', 'amazing', 'good', 'outstanding', 'recommend', 'happy', 'love'];
    const negativeWords = ['bad', 'terrible', 'awful', 'disappointed', 'not satisfied', 'poor', 'hate'];

    const textLower = text.toLowerCase();
    const positiveCount = positiveWords.filter((w) => textLower.includes(w)).length;
    const negativeCount = negativeWords.filter((w) => textLower.includes(w)).length;

    let sentiment: string;
    let confidence: number;

    if (positiveCount > negativeCount) {
      sentiment = 'positive';
      confidence = Math.min(0.5 + positiveCount * 0.1, 0.95);
    } else if (negativeCount > positiveCount) {
      sentiment = 'negative';
      confidence = Math.min(0.5 + negativeCount * 0.1, 0.95);
    } else {
      sentiment = 'neutral';
      confidence = 0.6;
    }

    return {
      sentiment,
      confidence: Math.round(confidence * 100) / 100,
      wordCount: text.split(/\s+/).length,
      positiveIndicators: positiveCount,
      negativeIndicators: negativeCount,
    };
  }

  private summarize(text: string): SummaryResult {
    const sentences = text.split('.').map((s) => s.trim()).filter((s) => s.length > 0);
    const summarySentences = sentences.length > 2 ? sentences.slice(0, 2) : sentences;
    const summary = summarySentences.length > 0 ? summarySentences.join('. ') + '.' : text;

    return {
      summary,
      originalLength: text.length,
      summaryLength: summary.length,
      compressionRatio: Math.round((summary.length / Math.max(text.length, 1)) * 100) / 100,
      sentenceCount: sentences.length,
    };
  }

  private detectTrends(data: Array<Record<string, unknown>>): TrendResult {
    if (!data || data.length < 2) {
      return { trend: 'insufficient_data', dataPoints: data?.length || 0, confidence: 0.0 };
    }

    const values = data.filter((d) => typeof d === 'object' && d !== null).map((d) => (d.value as number) || 0);

    if (values.length < 2) {
      return { trend: 'insufficient_data', dataPoints: values.length, confidence: 0.0 };
    }

    const halfIndex = Math.floor(values.length / 2);
    const firstHalfAvg = values.slice(0, halfIndex).reduce((a, b) => a + b, 0) / halfIndex;
    const secondHalfAvg = values.slice(halfIndex).reduce((a, b) => a + b, 0) / (values.length - halfIndex);
    const changePct = firstHalfAvg !== 0 ? ((secondHalfAvg - firstHalfAvg) / firstHalfAvg) * 100 : 0;

    let trend: string;
    if (changePct > 5) trend = 'upward';
    else if (changePct < -5) trend = 'downward';
    else trend = 'stable';

    return {
      trend,
      changePercent: Math.round(changePct * 10) / 10,
      dataPoints: values.length,
      startValue: values[0],
      endValue: values[values.length - 1],
      confidence: 0.78,
    };
  }

  async getTrustScore(): Promise<number> {
    try {
      const score = await this.a2a.getTrustScore(this.agentId);
      return score.score;
    } catch {
      return 0.0;
    }
  }
}
